### ISSUE TITLE
Strategy for maximizing number of HTTP requests

### ISSUE THREAD

**Issue by zdrd** [Norconex/collector-http#302]
I'm trying to crawl as many URLs as possible, with the goal being speed, and minimal idle time. I have experimented with many different configurations, such as having many threads (200-400) on a fairly powerful computer (~16 vCPUs, 50+gb ram), and also few threads (2-10 threads) on multiple smaller machines (5 single-core servers each running a completely separate http-collector simultaneously), and the issue I am having is that there doesn't seem to be any noticeable speed increase in running a http-collector on a small machine with few threads versus on a large machine with hundreds of threads. Is there any settings that you know of that I may be neglecting which may be contributing to this? My config has same-host timeout as `<delay default="1500"/>`. I considered that perhaps it is due to the list of seed-urls (`<startUrls>`) I provide, so I tried with both very large (hundreds of hosts (more hosts than threads)) and small, also no difference in speed. In theory, if I give different 400 hosts with 400 threads, and each took 1 second to reply, then I'd expect 400 urls downloaded per 1 second, but this isn't the type of behavior I am seeing. I notice from following along with the stdout that the collector outputs, it seems to idle on a single host for the courtesy timeout I have set to about 1.5 seconds, but if I have spawned 400 threads, would the crawler not use that 1.5 seconds to crawl another domain using one of the other threads? Hopefully it makes sense what I am asking/describing. Any thoughts or suggestions would help a lot, thanks!

**Comment by essiembre** [comment]
There might be other bottlenecks that are hare to troubleshoot given how many URLs/threads you are using.  Just to rule out the delay, have you tried setting it to zero to see if that makes any difference?  Have you disabled sitemap support? Because URLs are processed in the order they are discovered while crawling.  Usually when you have multiple starting URLs, requests get spread naturally between domains over time.  But if there are big sitemaps, you may have a big batch of URLs inserted at once that will take a while to process before a URL from another domain gets pulled form the queue. Where to you store the results? Can it be your target repository that has trouble keeping up?  Can it be your hardware that has trouble keeping up?  Document parsing can be quite intense on CPU for some document types.  Given you only have 16 CPUs for 400 threads, that seems a very high number of thread to use.  Are CPUs at maximum capacity when this happens?

Can't think of anything else right now.  If you are familiar with JVM profilers, maybe using one could tell you where bottlenecks are.

**Comment by zdrd** [comment]
I think the issue is probably being caused by the order of the queue reads, combined with the delay I have set. I think if maybe I altered the code slightly so that it does a random read from the queue, I would end up getting significantly more crawled pages per second. The odd thing is that I am able to successfully run up to 12,000 threads on a 16-core machine, but the CPU % never reaches even half its capacity no matter how many threads I use (it breaks at around 20k threads). I used `top` to check out the threads while the crawler is running, and every thread is marked as 'sleeping' except for 1 marked 'active', which seems strange to me. I'm going to take your advice and first run a java profiler and try and track down what's going on. Thanks for the suggestions
